scanpy/external/tl/_harmony_timeseries.py

```diff
@@ -0,0 +1,462 @@
+
+
+    Returns
+    -------
+    `harmony_adata` of :class:`~anndata.AnnData` of normalized data:
+
+    - `.obsm['layout']` force directed layout
+
+    - `.uns['aff']` affinity matrix
+
+    - `.uns['aug_aff']` augmented affinity matrix
+
+    - `.uns['timepoint_connections']` timepoint connections
+
+    Example
+    -------
+
+    >>> import scanpy.external as sce
+    >>> import scanpy as sc
+    >>> import os
+
+    There are two ways to load data used with harmony_timeseries. If you have `csv`
+    files, follow these steps. If you have `AnnData` sets, follow the next snippet.
+
+    *Load sample data from csv files*
+
+    A sample data is available `here <https://github.com/dpeerlab/Harmony/tree/master/data>`_.
+
+    To view the plots, it is recommended to run jupyter notebook
+
+    >>> # harmony_dir = os.path.expanduser('~/Harmony/data/')
+    >>> csv_files = [
+    ...     harmony_dir + i for i in [
+    ...         'Lib1-3_E3.5.csv',
+    ...         'Lib1-4_E3.5.csv',
+    ...         'Lib1-1_E4.5.csv',
+    ...         'Lib1-2_E4.5.csv'
+    ...     ]
+    ... ]
+    >>> sample_names = [
+    ...     'sa1_Rep1',
+    ...     'sa1_Rep2',
+    ...     'sa2_Rep1',
+    ...     'sa2_Rep2'
+    ... ]
+    >>> # timepoints = ['sa1', 'sa2', 'sa3', 'sa4']
+    >>> d = sce.tl.harmony_timeseries(data=csv_files, sample_names=sample_names)
+
+    At this point, a new class object, `d`, will be instantiated
+
+    *Loading sample data from `AnnData` sets*
+
+    Assuming you have 4 `AnnData` sets from 2 time points with 2 replicates,
+    instead of `csv` files, the following snippet shows how to feed a list of
+    `AnnData` sets to harmony_timeseries.
+
+    Loading sample data from Scanpy builtin datasets:
+
+    >>> adata_ref = sc.datasets.pbmc3k()
+
+    Splitting `adata_ref` into 4 `adata` sets (just to show how it works). Normally,
+    you would have different `AnnData` sets from different time points ready to use.
+
+    >>> n = int(adata_ref.shape[0] / 4)
+    >>> datalist = [adata_ref[n * i : n * (i + 1)] for i in range(4)]
+    >>> sample_names = [
+    ...     'sa1_Rep1',
+    ...     'sa1_Rep2',
+    ...     'sa2_Rep1',
+    ...     'sa2_Rep2'
+    ... ]
+    >>> # Instantiate harmony_timeseries:
+    >>> d = sce.tl.harmony_timeseries(data=datalist, sample_names=sample_names)
+
+    The next steps apply on both cases of loaded data.
+
+    **Pre-processing**
+
+    Data normalization is performed internally once `sce.tl.harmony_timeseries` is instantiated.
+
+    The data can be log transformed by setting `log_transform` to `True`
+
+    >>> d.log_transform = True
+
+    The created object `d.harmony_timeseries` can be used to override the default parameters as well
+    as running the different methods embedded in it for plotting
+
+    Follow the next step to pass the data to harmony methods, to generate the
+    return objects listed above.
+
+    **Run harmony_timeseries**
+
+    >>> d.process(n_components=902)
```

Comment:
In scanpy, tools run when you call `sc/sce.tl/pp.toolname(adata)`, they don’t return an object that isn’t an AnnData.